### PR TITLE
[TC] class_tactics register for custom class solver

### PR DESCRIPTION
**Pull request by FissoreD**
This adds a hook for type classes, so that plugins can provide solutions during type class inference.

### CODE REVIEW

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@FissoreD 
```
tactics/class_tactics.ml:1198: trailing whitespace.
+      try 
tactics/class_tactics.ml:1200: trailing whitespace.
+        if cond env evd s then solver else find_solver tl 
```
you can typically fix these by `git rebase master --whitespace=fix` and/or installing a commit hook that should be somewhere in `dev/`

**Comment by gares** [comment]
serapi is the only failure and is unrelated, so we can merge once the linter problem is fixed.

File: tactics/class_tactics.mli [hunk 1]
```diff
@@ -57,6 +57,22 @@ val autoapply : constr -> Hints.hint_db_name -> unit Proofview.tactic
...
 val resolve_tc : constr -> unit Proofview.tactic
 
+type solver = (Environ.env ->
+    Evd.evar_map ->
+    Constr.metavariable option ->
+    bool ->
+    best_effort:bool ->
+    (Evd.evar_map -> Evar.t -> bool) ->
+    (bool * Evd.evar_map) option)
+
```
Comment (gares) [inline tactics/class_tactics.mli:67]:
```suggestion
type solver = (Environ.env ->
    Evd.evar_map ->
    int option -> (* depth *)
    bool -> (* unique *)
    best_effort:bool ->
    (Evd.evar_map -> Evar.t -> bool) -> (* filter *)
    (bool * Evd.evar_map) option)

```
you could also use labels. You should also document the return type, what is the bool BTW?

**Comment by gares** [comment]
@coqbot run full ci

File: tactics/class_tactics.ml [hunk 2]
```diff
@@ -442,6 +442,17 @@ let make_hints env sigma (modes,st) only_classes sign =
       else hints)
     sign db
 
+module Intpart = Unionfind.Make(Evar.Set)(Evar.Map)
+
+type solver =
```
Comment (ppedrot) [inline tactics/class_tactics.ml:447]: Could we get a record wrapper here? A function with that many arguments is prone to generate unreadable error messages if we modify the API.

Comment (gares) [reply tactics/class_tactics.ml:447]: Ok, but that is the type of the built-in solver as well. You want just the public api to be wrapped, or the internal one as well?

Comment (gares) [reply tactics/class_tactics.ml:447]: And to help @FissoreD please give a pointer to a wrapped api, I don't know exactly myself what you want exactly :-)

Comment (ppedrot) [reply tactics/class_tactics.ml:447]: At least the public API.

Comment (ppedrot) [reply tactics/class_tactics.ml:447]: Just something along the lines of
```ocaml
type solver = { solver : 
  Environ.env ->
  Evd.evar_map ->
  depth:int option ->
  unique:bool ->
  best_effort:bool ->
  goals:Evar.t list ->
  (bool * Evd.evar_map)
}
```

Comment (FissoreD) [reply tactics/class_tactics.ml]: @ppedrot in my last commit I wrap the type solver in a record :+1:

**Comment by gares** [comment]
@coqbot run full ci

File: tactics/class_tactics.ml [hunk 3]
```diff
@@ -1227,18 +1230,18 @@ let resolve_all_evars depth unique env p oevd fail =
...
           match evars_to_goals p evd with
           | Some (goals, nongoals) ->
-            let solver = find_solver env evd comp in
-            let evm, goalsls = Search.preprocess_goals evd goals in
-            let finished, evm' = solver env evm ~goals:goalsls ~best_effort:true ~depth ~unique in
-            let evm' = Search.post_process_goals ~goals ~nongoals ~evm ~evm' ~finished in
-            if has_undefined p oevd evm' then
+            let { solver } = find_solver env evd comp in
+            let evd, old_metas, sorted_goals = Search.preprocess_goals evd goals in
+            let finished, evd = solver env evd ~goals:sorted_goals ~best_effort:true ~depth ~unique in
```
Comment (ppedrot) [inline tactics/class_tactics.ml:1235]: Last nitpick, I promise: for even more robustness you should perform the projection on this line, as
```ocaml
            let solver = find_solver env evd comp in
            let evd, old_metas, sorted_goals = Search.preprocess_goals evd goals in
            let finished, evd = solver.solver env evd ~goals:sorted_goals ~best_effort:true ~depth ~unique in
```

Comment (FissoreD) [reply tactics/class_tactics.ml]: No, problem :-)
I've commited right now what you suggest

**Comment by gares** [comment]
@coqbot run full ci

File: tactics/class_tactics.ml (hunk 3 above)
Comment (ppedrot) [reply tactics/class_tactics.ml]: I'm extremely wary of the mess introduced by optional and named arguments. Having a statically known arity at apply time is always good to drive the unification errors into a readable output. (Also there might be considerations about functions being fully applied.)

**Comment by gares** [comment]
@coqbot merge now

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**